perllib/FixMyStreet/App/Controller/Admin/Waste.pm

```diff
@@ -106,35 +106,48 @@ sub bulky_items : Chained('body') {
     if ($c->req->method eq 'POST') {
         $c->forward('/auth/check_csrf_token');
 
+        $c->stash->{errors} = 0;
+
         my @indices = grep { /^bartec_id\[\d+\]/ } keys %{ $c->req->params };
         @indices = map { /(\d+)/ } @indices;
         my @items;
         foreach my $i (@indices) {
-            my $bartec_id = $c->get_param("bartec_id[$i]");
-            next unless $bartec_id;
-            push @items, {
+            my $item = {
                 bartec_id => $c->get_param("bartec_id[$i]"),
                 category => $c->get_param("category[$i]"),
                 name => $c->get_param("name[$i]"),
                 message => $c->get_param("message[$i]"),
                 price => $c->get_param("price[$i]"),
             };
+            # OK to store errors in $item itself as it won't get persisted,
+            # and $i might not be the same when form is re-rendered.
+            foreach (qw(name category bartec_id)) {
+                if (!$item->{$_}) {
+                    $item->{errors} ||= {};
+                    $item->{errors}->{$_} = _("This field is required.");
+                    $c->stash->{errors} = 1;
+                }
```

Comment:
You may also want to limit the length of inputs such as 'message'; I can currently input vast swathes of text (not that it's likely for someone to do that in reality).